packages/flutter/lib/src/cupertino/colors.dart

```diff
@@ -80,3 +93,608 @@ class CupertinoColors {
+@immutable
+class CupertinoDynamicColor extends Color {
+  /// Creates a [Color] that can be resolved to different [Color]s in different
+  /// [BuildContext], using [CupertinoDynamicColor.resolve].
+  ///
+  /// [defaultColor] will be used in other colors' absence, and it must not be null,
+  /// unless all the other colors are not null.
+  CupertinoDynamicColor({
+    Color defaultColor,
+    Color normalColor,
+    Color darkColor,
+    Color highContrastColor,
+    Color darkHighContrastColor,
+    Color elevatedColor,
+    Color darkElevatedColor,
+    Color highContrastElevatedColor,
+    Color darkHighContrastElevatedColor,
+  }) : this._(
+         defaultColor,
+         <Color>[
+           normalColor,
+           darkColor,
+           highContrastColor,
+           darkHighContrastColor,
+           elevatedColor,
+           darkElevatedColor,
+           highContrastElevatedColor,
+           darkHighContrastElevatedColor,
+         ]
+       );
+
+  /// Creates a Color that can be resolved to different [Color]s, based on the given
+  /// [BuildContext]'s color vibrancy and accessibility contrast.
+  ///
+  /// [defaultColor] will be used in other colors' absence, and it must not be null,
+  /// unless all the other colors are not null.
+  CupertinoDynamicColor.withVibrancyAndContrast({
+    Color defaultColor,
+    Color normalColor,
+    Color darkColor,
+    Color highContrastColor,
+    Color darkHighContrastColor,
+  }) : this(
+    defaultColor: defaultColor,
+    normalColor: normalColor,
+    darkColor: darkColor,
+    highContrastColor: highContrastColor,
+    darkHighContrastColor: darkHighContrastColor,
+    elevatedColor: normalColor,
+    darkElevatedColor: darkColor,
+    highContrastElevatedColor: highContrastColor,
+    darkHighContrastElevatedColor: darkHighContrastColor,
+  );
+
+  /// Creates a Color that can be resolved to different [Color]s, based on the given
+  /// [BuildContext]'s color vibrancy.
+  ///
+  /// [defaultColor] will be used in other colors' absence, and it must not be null,
+  /// unless all the other colors are not null.
+  CupertinoDynamicColor.withVibrancy({
+    Color defaultColor,
+    Color normalColor,
+    Color darkColor,
+  }) : this(
+    defaultColor: defaultColor,
+    normalColor: normalColor,
+    darkColor: darkColor,
+    highContrastColor: normalColor,
+    darkHighContrastColor: darkColor,
+    elevatedColor: normalColor,
+    darkElevatedColor: darkColor,
+    highContrastElevatedColor: normalColor,
+    darkHighContrastElevatedColor: darkColor,
+  );
+
+  CupertinoDynamicColor._(
+    this.defaultColor,
+    this._colorMap,
+  ) : assert(defaultColor != null || !_colorMap.contains(null)),
+      assert(_colorMap.length == 8),
+      super(defaultColor?.value ?? _colorMap[0].value);
+
+  /// The defaultColor color to use when the requested color is not specified.
+  ///
+  /// Must not be null unless all other colors are specified.
+  final Color defaultColor;
```

Comment:
I don't think this is really needed, the CupertinoDynamicColor _is_ the default color. If we always assume that all 8 colors are defined, why do need this? We should talk this over.